algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/molecular/HasMtapDeletion.kt

```diff
@@ -0,0 +1,62 @@
+package com.hartwig.actin.algo.evaluation.molecular
+
+import com.hartwig.actin.algo.evaluation.EvaluationFactory
+import com.hartwig.actin.algo.evaluation.EvaluationFunction
+import com.hartwig.actin.algo.evaluation.IhcTestEvaluation
+import com.hartwig.actin.datamodel.PatientRecord
+import com.hartwig.actin.datamodel.algo.Evaluation
+import com.hartwig.actin.datamodel.algo.EvaluationResult
+import com.hartwig.actin.datamodel.molecular.MolecularTestTarget
+
+private const val MTAP = "MTAP"
+private const val CDKN2A = "CDKN2A"
+
+class HasMtapDeletion : EvaluationFunction {
+
+    override fun evaluate(record: PatientRecord): Evaluation {
+        val mtapTested =
+            record.molecularTests.any { test -> test.targetSpecification?.testsGene(MTAP) { it == listOf(MolecularTestTarget.DELETION) } == true }
+                    || IhcTestEvaluation.create(MTAP, record.ihcTests).filteredTests.isNotEmpty()
+
+        return when {
+            mtapTested -> {
+                val mtapDeletion = GeneIsInactivated(MTAP, onlyDeletions = true).evaluate(record)
+                val mtapIhcLoss = ProteinIsLostByIhc(MTAP).evaluate(record)
+                val mtapInactivation = GeneIsInactivated(MTAP, onlyDeletions = false).evaluate(record)
+
+                when {
+                    isPassOrWarn(mtapIhcLoss) -> mtapIhcLoss
+
+                    isPassOrWarn(mtapDeletion) -> mtapDeletion
+
+                    isPassOrWarn(mtapInactivation) -> {
+                        mtapInactivation.copy(
+                            result = EvaluationResult.WARN,
+                            passMessages = emptySet(),
+                            warnMessages = mtapInactivation.passMessages.ifEmpty { mtapInactivation.warnMessages }
+                        )
+                    }
+
+                    else -> EvaluationFactory.fail("No $MTAP deletion")
+                }
```

Comment:
Just an idea, but if you implemented `Evaluation.takeIfPassOrWarn()`, you could rewrite this as:
```suggestion
                ProteinIsLostByIhc(MTAP).evaluate(record).takeIfPassOrWarn()
                    ?: GeneIsInactivated(MTAP, onlyDeletions = true).evaluate(record).takeIfPassOrWarn()
                    ?: GeneIsInactivated(MTAP, onlyDeletions = false).evaluate(record).takeIfPassOrWarn()?.copy(
                        result = EvaluationResult.WARN,
                        passMessages = emptySet(),
                        warnMessages = mtapInactivation.passMessages.ifEmpty { mtapInactivation.warnMessages }
                    )
                    ?: EvaluationFactory.fail("No $MTAP deletion")
```

Reply:
No, it doesn't exist already.

You could argue about readability here, but I think there's actually an improvement in inlining the evaluations so you don't have to track 3 additional `val`s to see what's being referenced on each line. `takeIf` is also nice when your condition looks like `if (f(x)) return x else y` because you can do it without referencing `x` twice (`x.takeIf(::f) ?: y`). And sometimes I rationalize using slick Kotlin syntax just by citing the rule of cool 😎 

Do either of you think `takeIf` hurts readability?

Reply:
I like it! It might be slightly less straightforward to read but still easy to follow and since this is way more concise I'm all in favour of using `takeIf`